src/mds/Locker.cc

```diff
@@ -379,6 +387,28 @@ bool Locker::acquire_locks(const MDRequestRef& mdr,
       ceph_assert(0 == "locker unknown lock operation");
     }
   }
+  bool injected_quiesce_lock = false;
+  for (auto& [co, flags]: quiesce_op_flags) {
+    CInode* in = static_cast<CInode*>(co);
+    if (flags) {
+      bool exclusive = (flags & LockOp::XLOCK);
+      dout(20) << "will take " << (exclusive ? "an exclusive" : "a shared") 
+               << " quiesce lock for inode " << *in << " mdr: " << *mdr << dendl;
+      if (exclusive) {
+        lov.add_xlock(&in->quiescelock, 0);
```

Comment:
This can lead to an unexpected situation where you have something like:

```
[x iquiesce ino=0xabc, x iquiesce ino=0xdef, w ilinklock ino=0xabc, w ilinklock ino=0def]
```

even after sorting because `sort_and_merge` does not sort by the lock parent (`ino` in this example).

I don't **think** this is a problem and probably deliberate on your part but there should be a comment explaining that quiescelocks are taken first no matter what else is in the vector (and, note: the quiescelocks may be taken out of order). The reason this isn't a problem is because if we cannot acquire a quiescelock then we drop everything.

I realize this PR is time sensitive so please just followup with a cleanup PR after this is merged.

Reply:
Hmm, it's a point I haven't given it a lot of thought. As for taking locks in the correct order per directory structure, we should rely here on the caller. It means that we should use a structure that preserves the order of objects in the quiesce list.

Thanks for the comment, I opened https://tracker.ceph.com/issues/66094